### PR TITLE
Addition of setting for whitelisting of IP address ranges

### PR DESCRIPTION
**Pull request by Cimbali**
## Description

Adds a setting to allow whitelisting of IP address ranges, to extend split-tunnelling capabilities.

Unfortunately I haven’t yet figured out how to efficiently modify the qml to add a settings view for it, so this is only the function bit so far.

In the mean time any feedback would be helpful. I use this as a patch and edit the json settings directly.

## Reference

As described in #5182 

## Checklist
    
- [x] My code follows the style guidelines for this project
        I’ve ran clang-format-11.0.1
- [x] I have not added any packages that contain high risk or unknown licenses (GPL,  LGPL, MPL, etc. consult with DevOps if in question)
- [x] I have performed a self review of my own code
- [x] I have commented my code PARTICULARLY in hard to understand areas
- [x] I have added thorough tests where needed
       “Where needed” is quite unclear but there seem to not be much testing for similar settings?

### CODE REVIEW

**Comment by oskirby** [comment]
I think for naming conventions, these settings should be named to something along the lines of "excludedAddresses" because we are excluding these addresses from being routed into the VPN. The term "allowed addresses" in the wireguard world specifically means a range that must be routed into the VPN tunnel.

**Comment by Cimbali** [comment]
I’ll gladly rename to excluded (rather than allowed) IPs, and add it to the CLI.

**Comment by Cimbali** [comment]
Here’s what playing with the proposed CLI looks like (using random addresses):
```
> mozillavpn excludeip list
Excluded IP v4 addresses:
Excluded IP v6 addresses:
> mozillavpn excludeip add 2001:DB8::00/100
> mozillavpn excludeip list
Excluded IP v4 addresses:
Excluded IP v6 addresses:
- 2001:db8::/100
> mozillavpn excludeip add 9.9.9.9
> mozillavpn excludeip add 192.168.0.0/16
> mozillavpn excludeip list
Excluded IP v4 addresses:
- 9.9.9.9/32
- 192.168.0.0/16
Excluded IP v6 addresses:
- 2001:db8::/100
> mozillavpn excludeip remove 192.168.0.0/255.255.0.0
> mozillavpn excludeip list
Excluded IP v4 addresses:
- 9.9.9.9/32
Excluded IP v6 addresses:
- 2001:db8::/100
> mozillavpn excludeip remove dead::beef
dead::beef/128 is not currently an excluded IP address.
```

File: src/commands/commandexcludeip.cpp [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    int command = -1;
+    if (tokens.length() > 1) {
+      command = commandList.indexOf(tokens[1]);
+    }
+
+    QTextStream stream(stdout);
+    if (command < 0) {
+      stream << "usage: " << tokens[0] << " <list|add|remove> [ip]" << Qt::endl;
+      return 1;
+    } else if (command == LIST && tokens.length() != 2) {
```
Comment (bakulf) [inline src/commands/commandexcludeip.cpp:39]: No `else` after a `return`.

File: src/commands/commandexcludeip.cpp [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+      stream << "usage: " << tokens[0] << " " << tokens[1] << Qt::endl;
+      return 1;
+    } else if (command != LIST && tokens.length() != 3) {
```
Comment (bakulf) [inline src/commands/commandexcludeip.cpp:42]: Same here

File: src/commands/commandexcludeip.h [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef COMMANDDEVICE_H
+#define COMMANDDEVICE_H
+
+#include "command.h"
+
+class CommandExcludeIP final : public Command {
+ public:
+  CommandExcludeIP(QObject* parent);
```
Comment (bakulf) [inline src/commands/commandexcludeip.h:12]: Explicit

**Review by bakulf — APPROVED** [review]

**Review by oskirby — APPROVED** [review]

**Comment by bakulf** [comment]
@Cimbali we are changing the entire structure or the repo. Can I pick your PR, rebase and push it again?

**Comment by Cimbali** [comment]
Yes of course @bakulf

**Comment by Cimbali** [comment]
I will also have time to rebase this week @bakulf  if you prefer.

**Comment by oskirby** [comment]
@Cimbali, I'm afraid @bakulf is out of the office for the next couple of weeks. I would be happy to merge this PR if you want to rebase it in the meantime.

**Comment by Cimbali** [comment]
Done and tested.

Sorry for the noise I somehow can’t seem to figure out how to generate the glean filles locally… Anyways I’ve downloaded the CI source bundles and compiled from that, and everything’s behaving as intended @oskirby

**Comment by Cimbali** [comment]
Actually, strangely enough, activating the tunnel hangs when there are 6 or more excluded IP ranges…

**Comment by birdsarah** [comment]
@Cimbali your PR shouldn't include changes to i18n...then we can merge!

**Comment by Cimbali** [comment]
Fixed the hang, and added a test for it. This PR is ready for review/merge again.

**Review by bakulf — APPROVED** [review]

**Pull request merged**